### PR TITLE
add script to support XLIFF Translation

### PR DESCRIPTION
**Pull request by hchoura**

### CODE REVIEW

File: src/foam/nanos/controller/ApplicationController.js [hunk 1]
```diff
@@ -395,6 +397,41 @@ foam.CLASS({
...
+                console.log(ea.source)
+                console.log(err)
+              }
+              if (!!s) {
+                s[ea.source.substring(ea.source.lastIndexOf('.')+1)] = ea.target;
+              }
+            })
+        })
+      } catch (err) {//TODO
+        this.notify(this.LANGUAGE_FETCH_ERR, 'error');
```
Comment (jlhughes) [inline src/foam/nanos/controller/ApplicationController.js:448]: @hchoura can you create a ticket to capture this scenario - we need to 'notify' ourselves - send the error back to our noc that this situation is occurring.

File: src/services [hunk 2]
```diff
@@ -1069,3 +1069,20 @@ p({
     }
   """
 })
+
+p({
+  "class":"foam.nanos.boot.NSpec",
+  "name":"XLIFFTranslationDAO",
+  "serve":true,
+  "serviceScript":"return new foam.dao.EasyDAO.Builder(x).setAuthorize(false).setJournalName(\"XLIFFTranslations\").setJournalType(foam.dao.JournalType.SINGLE_JOURNAL).setOf(foam.i18n.XLIFFTranslationValue.getOwnClassInfo()).build();",
```
Comment (jlhughes) [inline src/services]: please break this into multiple lines, using the """ format.  Helps with future commits.

File: src/services [hunk 3]
```diff
@@ -1069,3 +1069,20 @@ p({
...
+  "client":"{\"of\":\"foam.i18n.XLIFFTranslationValue\"}"
+})
+
+p({
+  "class":"foam.nanos.boot.NSpec",
+  "name":"LocaleDAO",
+  "description":"Locale DAO",
+  "serve":true,
+  "serviceScript":"return new foam.dao.EasyDAO.Builder(x).setAuthorize(false).setJournalName(\"locales\").setJournalType(foam.dao.JournalType.SINGLE_JOURNAL).setOf(foam.i18n.Locale.getOwnClassInfo()).build();",
```
Comment (jlhughes) [inline src/services]: Same here.

File: src/scripts [hunk 4]
```diff
@@ -59,4 +59,248 @@ p({
       print(t.getMessage());
     }
   """
-})
\ No newline at end of file
+})
+
+p({
```
Comment (jlhughes) [inline src/scripts]: Perhaps move these test scripts into 
src/foam/i18n/scripts.jrl 

Also, where are the scripts to extract all the messages and generate the xliff file? 
Assume it's similar to these without the locale, variant, and target.

File: src/services [hunk 5]
```diff
@@ -1069,3 +1069,46 @@ p({
     }
   """
 })
+
+p({
```
Comment (jlhughes) [inline src/services]: same, move these into src/foam/i18n/services.jrl
Reduces code merge conflicts in the future.

File: src/foam/i18n/services.jrl [hunk 6]
```diff
@@ -0,0 +1,42 @@
+p({
+  "class":"foam.nanos.boot.NSpec",
+  "name":"XLIFFTranslationDAO",
```
Comment (jlhughes) [inline src/foam/i18n/services.jrl]:
```suggestion
  "name":"xliffTranslationDAO",
```

File: src/foam/i18n/services.jrl [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  """
+    {
+      "of":"foam.i18n.XLIFFTranslationValue"
+    }
+  """
+})
+
+p({
+  "class":"foam.nanos.boot.NSpec",
+  "name":"LocaleDAO",
```
Comment (jlhughes) [inline src/foam/i18n/services.jrl]:
```suggestion
  "name":"localeDAO",
```

File: src/foam/nanos/controller/ApplicationController.js [hunk 8]
```diff
@@ -395,6 +397,49 @@ foam.CLASS({
...
     },
 
+    async function fetchLanguage() {
+      try {
+        let l = localStorage.getItem('localeLanguage');
+        if ( l !== undefined ) foam.locale = l;
+        //TODO manage more complicated language. 'en-CA'
+        if ( foam.locale !== 'en' && foam.locale !== 'en-US' ) {
+          let ctx = this.__subContext__;
+          let d = await  this.__subContext__.LocaleDAO;
```
Comment (jlhughes) [inline src/foam/nanos/controller/ApplicationController.js]:
```suggestion
          let d = await  this.__subContext__.localeDAO;
```

File: src/foam/u2/Element.js [hunk 9]
```diff
@@ -1627,7 +1631,24 @@ foam.CLASS({
         } else if ( typeof c === 'function' ) {
           throw new Error('Unsupported');
         } else {
-          es.push(c);
+          //foam.locale = 'fr';
+          if ( typeof c === 'object' && c.data !== undefined && c.data.id !== undefined ) {
+            var self = this;
+            var expr = foam.mlang.Expressions.create();
+            let d =  this.__subContext__.LocaleDAO;
```
Comment (jlhughes) [inline src/foam/u2/Element.js]:
```suggestion
            let d =  this.__subContext__.localeDAO;
```

File: src/foam/i18n/services.jrl [hunk 10]
```diff
@@ -0,0 +1,42 @@
+p({
+  "class":"foam.nanos.boot.NSpec",
+  "name":"XLIFFTranslationDAO",
+  "serve":true,
+  "serviceScript":
+  """
+    return new foam.dao.EasyDAO.Builder(x)
+      .setAuthorize(false)
+      .setJournalName("XLIFFTranslations")
```
Comment (jlhughes) [inline src/foam/i18n/services.jrl]:
```suggestion
      .setJournalName("xliffTranslations")
```

File: src/foam/i18n/scripts.jrl [hunk 11]
```diff
@@ -0,0 +1,243 @@
+p({
```
Comment (jlhughes) [inline src/foam/i18n/scripts.jrl:1]: Is this a test script? Asking as I don't understand the the locale, variant, ... settings in the forEach.

Comment (hchoura) [reply src/foam/i18n/scripts.jrl:1]: Yes, it is for test.
But we can change it in the production.

File: src/foam/u2/Element.js [hunk 12]
```diff
@@ -1627,7 +1631,24 @@ foam.CLASS({
         } else if ( typeof c === 'function' ) {
           throw new Error('Unsupported');
         } else {
-          es.push(c);
+          //foam.locale = 'fr';
```
Comment (jlhughes) [inline src/foam/u2/Element.js]:
```suggestion
```

**Review by jlhughes — APPROVED** [review]

**Pull request merged**